[Flang] Implement RENAME intrinsic (code-gen + runtime entry point)

mjklemm opened the pull request:
This PR implements the RENAME intrinsic, which is a GFortran extension (see https://gcc.gnu.org/onlinedocs/gfortran/intrinsic-procedures/rename.html).

mjklemm committed:
Implement code-gen for RENAME intrinsic

mjklemm committed:
Add runtime implementation for non-Windows systems

mjklemm committed:
Also use the POSIX API in Windows

mjklemm committed:
Format source code

mjklemm committed:
Add test for RENAME code-gen

mjklemm committed:
Update flang/docs/Intrinsics.md

mjklemm committed:
Fix namelist.f90 test to not use RENAME

tblah commented on flang/runtime/misc-intrinsic.cpp:
```diff
@@ -55,6 +55,30 @@ static RT_API_ATTRS void TransferImpl(Descriptor &result,
+  // We simply call rename(2) from POSIX
+  int result = rename(pathSrc, pathDst);
+  if (status) {
+    StoreIntToDescriptor(status, result, terminator);
```
Just a suggestion, feel free to ignore:

If the result is non-zero it would be more useful to return `errno` to indicate what went wrong instead of just `-1` (glibc) or an implementation defined non-zero value (Windows)

mjklemm replied:
I have made that change also.  This is also what GFortran seems to do.

tblah commented on flang/runtime/misc-intrinsic.cpp:
nit: please include cstdio explicitly for the declaration of `rename`

mjklemm replied:
Will do.

tblah commented on flang/lib/Optimizer/Builder/IntrinsicCall.cpp:
```diff
@@ -5911,6 +5917,37 @@ IntrinsicLibrary::genReduce(mlir::Type resultType,
+    mlir::Value statusBox =
+        isStaticallyPresent(status)
+            ? fir::getBase(status)
+            : builder.create<fir::AbsentOp>(loc, boxNoneTy).getResult();
```
Earlier you defined the intrinsic as supporting a dynamically optional argument. To support this you will also need a runtime check for something like this

```
subroutine doRename(status)
  integer, optional :: status
  call rename("old", "new", status)
end subroutine
```

Here we cannot determine statically whether `status` is present, but if it is present it should be used. I think the current code would treat it as absent. For example, see the implementation of C_ASSOCIATED in this file.

mjklemm replied:
I have this example now:

```Fortran
module ren                                                                         
    contains                                                                       
        subroutine rename_opt(src, dst, status)                                    
            implicit none                                                          
            character(*) :: src                                                    
            character(*) :: dst                                                    
            integer, optional :: status                                            
                                                                                   
            call rename(src, dst, status)                                          
        end subroutine rename_opt                                                  
end module ren                                                                     
```

The `isStaticallyPresent` still says that `status` is present (which I what I would expect, too, as the argument is there in the code).  So, it seems like the handling of `OPTIONAL` seems to be done at the call site.  Would be an accurate assumption to make for the code-gen in this PR?

mjklemm replied:
Here's a bit of MLIR that shows what's going on:

```
    %5:2 = hlfir.declare %arg2 dummy_scope %0 {fortran_attrs = #fir.var_attrs<optional>, uniq_name = "_QMrenFren>
    %6 = fir.is_present %5#0 : (!fir.ref<i32>) -> i1
    %9 = fir.embox %5#1 : (!fir.ref<i32>) -> !fir.box<i32>
    %10 = fir.absent !fir.box<i32>
    %11 = arith.select %6, %9, %10 : !fir.box<i32>
```

This looks like the code-gen for `OPTIONAL` is already doing something sensible, right?

tblah replied:
LGTM, sorry for the confusion

tblah reviewed (commented):
Thank you for adding this

mjklemm committed:
Report same error code as GFortran

mjklemm committed:
Fix a few typos

tblah commented on flang/runtime/misc-intrinsic.cpp:
```diff
@@ -67,7 +68,13 @@ void RTDECL(Rename)(const Descriptor &path1, const Descriptor &path2,
-    StoreIntToDescriptor(status, result, terminator);
+    // When an error has happened,
+    int errorCode = 0; // Assume success
+    if (result < 0) {
```
nit: The Windows documentation just says "a non-zero value"
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/rename-wrename

mjklemm replied:
Good catch!  I have changed that.  Windows effectively reports -1 (which is in line with the docs :-)), but I have made it such that it works in any case.

tblah reviewed (approved):
Thanks for the update! LGTM

mjklemm committed:
Make this piece work on both Windows and Linux

Windows says it only reports a non-zero error code.  Effectively,
it is -1, but let's rather be save than sorry.

klausler commented on flang/runtime/misc-intrinsic.cpp:
```diff
@@ -55,6 +56,36 @@ static RT_API_ATTRS void TransferImpl(Descriptor &result,
+      path2.OffsetElement(), path2.ElementBytes(), terminator)};
+
+  // We simply call rename(2) from POSIX
+  int result = rename(pathSrc, pathDst);
```
use braced initialization here and below in the runtime

mjklemm replied:
Done.

klausler commented on flang/lib/Evaluate/intrinsics.cpp:
```diff
@@ -2610,7 +2620,7 @@ bool IntrinsicProcTable::Implementation::IsDualIntrinsic(
   // Collection for some intrinsics with function and subroutine form,
   // in order to pass the semantic check.
-  static const std::string dualIntrinsic[]{{"etime"}, {"getcwd"}};
+  static const std::string dualIntrinsic[]{{"etime"}, {"getcwd"}, {"rename"}};
```
As long as you're changing this line, consider using `{"etime"s, "getcwd"s, "rename"s};` instead.

mjklemm replied:
Done.

mjklemm committed:
Incorporate feedback by @klausler

vzakhari commented on flang/runtime/misc-intrinsic.cpp:
```diff
@@ -55,6 +56,36 @@ static RT_API_ATTRS void TransferImpl(Descriptor &result,
 extern "C" {
 RT_EXT_API_GROUP_BEGIN
 
+void RTDECL(Rename)(const Descriptor &path1, const Descriptor &path2,
```
`RTDECL` -> `RTDEF`

I suggest using `#if !defined(RT_DEVICE_COMPILATION)` to guard the body of the function, as long as the device runtimes do not provide POSIX `rename`.  I would also suggest issuing an `unsupported` crash message in the else clause of the check.

mjklemm replied:
I have created https://github.com/llvm/llvm-project/pull/99445 in response to this.  Please have a look there and kindly provide a review there.
